### PR TITLE
[Bugfix] Make Differentiate const

### PR DESCRIPTION
**Pull request by matthew-mccall**
See title. This should fix compile issues with Oasis::Extra on Linux+GCC

### CODE REVIEW

File: tests/UnaryExpressionTests.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        Oasis::Multiply {
+            Oasis::Real { 2.0 },
+            Oasis::Negate { Oasis::Variable { "x" } } },
+        Oasis::Multiply{
+            Oasis::Real { 3.0 },
+            Oasis::Variable { "x" } }
+    }; // 2x+3x
+
+    auto after = before.Substitute(Oasis::Variable { "x" }, Oasis::Real { 4.0 }); // after should some std::unique_ptr<Expression> such that it equals 2(4) + 3(4)
+    Oasis::Real four { 4 };
```
Comment (404Naz) [inline tests/UnaryExpressionTests.cpp:24]: you say you want it to equal 2(4)+3(4) but then check if it is equal to 4? Am I misunderstanding something?

Comment (404Naz) [reply tests/UnaryExpressionTests.cpp:24]: The comment should be -2(4) but I think the test itself is fine.

**Review by 404Naz — APPROVED** [review]
A comment is incorrect but I think everything code-wise is fine.

File: tests/UnaryExpressionTests.cpp (hunk 1 above)
Comment (matthew-mccall) [reply tests/UnaryExpressionTests.cpp:24]: You're right, the comment is incorrect

**Review by 404Naz — APPROVED** [review]
looks good.

**Pull request merged**